File: extensions/osint-tools/src/forensics/tracker.ts
import * as crypto from 'crypto';

export interface EvidenceRecord {
  id: string; // Format: EV-YYYYMMDD-XXXX
  timestamp: string; // ISO 8601 UTC
  toolName: string;
  sourceUrl: string;
  rawInput: any;
  rawOutput: string;
  sha256: string;
}

export class ForensicEvidenceTracker {
  private evidenceLog: EvidenceRecord[] = [];
  private chainOfCustody: string[] = [];

  constructor(private investigationId: string) {
    this.logChain("Initialisation de l'investigation médico-légale numérique.");
  }

  public logChain(action: string) {
    const time = new Date().toISOString();
    this.chainOfCustody.push(`[${time}] ${action}`);
  }

  public registerEvidence(toolName: string, url: string, input: any, output: any): EvidenceRecord {
    const rawOutputStr = typeof output === 'string' ? output : JSON.stringify(output);
    const sha256 = crypto.createHash('sha256').update(rawOutputStr).digest('hex');
    const id = `EV-${new Date().toISOString().replace(/[-:T]/g, '').slice(0, 8)}-${crypto.randomBytes(3).toString('hex')}`;

    const record: EvidenceRecord = {
      id,
      timestamp: new Date().toISOString(),
      toolName,
      sourceUrl: url,
      rawInput: input,
      rawOutput: rawOutputStr,
      sha256
    };

    this.evidenceLog.push(record);
    this.logChain(`Preuve collectée par l'outil "${toolName}" (ID: ${id}, SHA-256: ${sha256.slice(0, 16)}...)`);
    return record;
  }

  public getManifest() {
    return {
      investigationId: this.investigationId,
      generatedAt: new Date().toISOString(),
      evidenceCount: this.evidenceLog.length,
      evidence: this.evidenceLog,
      chainOfCustody: this.chainOfCustody
    };
  }
}

File: backend/app/forensics/tracker.py
import json
import hashlib
from datetime import datetime
import secrets

class ForensicEvidenceTracker:
    def __init__(self, investigation_id: str):
        self.investigation_id = investigation_id
        self.evidence_log = []
        self.chain_of_custody = []
        self.log_chain("Initialisation de l'investigation médico-légale numérique.")

    def log_chain(self, action: str):
        timestamp = datetime.utcnow().isoformat() + "Z"
        self.chain_of_custody.append(f"[{timestamp}] {action}")

    def register_evidence(self, tool_name: str, url: str, input_params: any, raw_output: any) -> dict:
        timestamp = datetime.utcnow().isoformat() + "Z"
        
        # Normaliser le format de l'output brut pour le hachage
        if isinstance(raw_output, (dict, list)):
            raw_output_str = json.dumps(raw_output, sort_keys=True)
        else:
            raw_output_str = str(raw_output)
            
        sha256 = hashlib.sha256(raw_output_str.encode("utf-8")).hexdigest()
        
        # Générer un ID unique de preuve EV-YYYYMMDD-HEX
        date_str = datetime.utcnow().strftime("%Y%m%d")
        random_hex = secrets.token_hex(3)
        evidence_id = f"EV-{date_str}-{random_hex.upper()}"

        record = {
            "id": evidence_id,
            "timestamp": timestamp,
            "tool_name": tool_name,
            "source_url": url,
            "raw_input": input_params,
            "raw_output": raw_output_str,
            "sha256": sha256
        }

        self.evidence_log.append(record)
        self.log_chain(f"Preuve collectée par l'outil '{tool_name}' (ID: {evidence_id}, SHA-256: {sha256[:16]}...)")
        return record

    def get_manifest(self) -> dict:
        return {
            "investigation_id": self.investigation_id,
            "generated_at": datetime.utcnow().isoformat() + "Z",
            "evidence_count": len(self.evidence_log),
            "evidence": self.evidence_log,
            "chain_of_custody": self.chain_of_custody
        }
